lib/sensors/temperature.cpp

```diff
@@ -2,81 +2,92 @@
     : logger_(logger),
       i2c_(i2c),
-      channel_(channel)
+      channel_(channel),
+      device_address_(device_address)
 {
 }
 
 Temperature::~Temperature()
 {
 }
 
-std::optional<std::int16_t> Temperature::read()
+std::optional<core::Result> Temperature::checkStatus()
 {
-  const auto status_check_result = i2c_->readByte(kTemperatureDefaultAddress, kStatus);
+  const auto status_check_result = i2c_->readByte(device_address_, kStatus);
   if (!status_check_result) {
     logger_.log(
       core::LogLevel::kFatal, "Failed to read temperature sensor status at channel %d", channel_);
     return std::nullopt;
   }
-  // TODOLater: change this from a fatal error to a debug and re-prompt read (with ROS)
-  if (status_check_result.value() == kBusy) {
-    logger_.log(core::LogLevel::kFatal,
+  if (*status_check_result == kBusy) {
+    logger_.log(core::LogLevel::kWarn,
                 "Failed to read, temperature sensor is not ready to be read at channel %d",
                 channel_);
-    return std::nullopt;
+    return core::Result::kFailure;
   }
-  if (status_check_result.value() == kTemperatureOverUpperLimit) {
+  if (*status_check_result == kTemperatureOverUpperLimit) {
     logger_.log(core::LogLevel::kFatal,
                 "Failed to read, temperature is above upper limit at channel %d",
                 channel_);
     return std::nullopt;
   }
-  if (status_check_result.value() == kTemperatureUnderLowerLimit) {
+  if (*status_check_result == kTemperatureUnderLowerLimit) {
     logger_.log(core::LogLevel::kFatal,
                 "Failed to read, temperature is below lower limit at channel %d",
                 channel_);
     return std::nullopt;
   }
-  const auto temperature_high_byte
-    = i2c_->readByte(kTemperatureDefaultAddress, kDataTemperatureHigh);
+  return core::Result::kSuccess;
+}
+
+std::optional<std::int16_t> Temperature::read()
+{
+  const auto temperature_high_byte = i2c_->readByte(device_address_, kDataTemperatureHigh);
   if (!temperature_high_byte) {
     logger_.log(
       core::LogLevel::kFatal, "Failed to read high byte for temperature at channel %d", channel_);
     return std::nullopt;
   }
-  const auto temperature_low_byte = i2c_->readByte(kTemperatureDefaultAddress, kDataTemperatureLow);
+  const auto temperature_low_byte = i2c_->readByte(device_address_, kDataTemperatureLow);
   if (!temperature_low_byte) {
     logger_.log(
       core::LogLevel::kFatal, "Failed to read low byte for temperature at channel %d", channel_);
     return std::nullopt;
   }
-  const std::int16_t temperature
-    = ((temperature_high_byte.value() << 8) | temperature_low_byte.value());
+  const std::int16_t temperature = ((*temperature_high_byte << 8) | *temperature_low_byte);
   logger_.log(
     core::LogLevel::kDebug, "Successfully read from temperature sensor at channel %d", channel_);
   // Scaling temperature as per the datasheet
   return static_cast<std::int16_t>(temperature * kTemperatureScaleFactor);
```

Comment:
temperature is already int16_t, think you only need to cast scaling factor.

Reply:
scale factor is a float so we need to static_cast the product else itll be a float

Reply:
![image](https://user-images.githubusercontent.com/58626285/218270104-ea0b2d5e-e82d-466f-a542-b0a63200f03d.png)
actually this is quite interesting coz it makes more sense to go with the more limiting variables (in terms of space), but c++ tries to go with more information over less.

Reply:
I'm chatting nonsense, pls ignore me!